packages/trace-viewer/src/ui/shared/dialog.tsx

```diff
@@ -0,0 +1,162 @@
+/*
+  Copyright (c) Microsoft Corporation.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+import * as React from 'react';
+
+export interface DialogProps {
+  className?: string;
+
+  open: boolean;
+  width: number;
+
+  verticalOffset?: number;
+
+  requestClose?: () => void;
+
+  hostingElement?: React.RefObject<HTMLElement>;
+}
+
+/**
+ * A dropdown dialog, positioned below the `hostingElement`. Uses a HTML `dialog` element
+ */
+export const Dialog: React.FC<React.PropsWithChildren<DialogProps>> = ({
+  className,
+  open,
+  width,
+  verticalOffset,
+  requestClose,
+  hostingElement,
+  children,
+}) => {
+  const dialogRef = React.useRef<HTMLDialogElement>(null);
+
+  // Allow window dimension changes to force a rerender
```

Comment:
We have a high threshold for comments in Playwright. Comments mean:
- "something very unusual is taking place, but please don't remove the seemingly wrong code before you understand the context"
- a sequence of complex actions is stacked in a single method and needs separation.